File: app-mobile/src/api/house.ts
/** Touse/Bouse weekly house leaderboard — mirrors routers/house.py (board card c175). */

import { request } from "./client";

export interface HouseBallot {
  campus_id: string;
  week_start: string;
  touse_chapter_id: string;
  bouse_chapter_id: string | null;
  updated_at: string;
}

export interface HouseStanding {
  chapter_id: string;
  org_name: string;
  chapter_name: string | null;
  rank: number;
  touse_votes: number;
  bouse_votes: number;
  net: number;
}

/**
 * A house that has not cleared the vote threshold this week.
 *
 * Kept separate from the ranking rather than sorted to the bottom, and the screen must
 * keep it that way: "came last" and "three people voted" are different statements about
 * a real organisation, and the bottom of this leaderboard is public.
 */
export interface UnrankedHouse {
  chapter_id: string;
  org_name: string;
  chapter_name: string | null;
  votes: number;
}

export interface HouseLeaderboard {
  campus_id: string;
  week_start: string;
  /** The sample every number here is drawn from. Always shown, never implied. */
  ballots_cast: number;
  min_votes_to_rank: number;
  ranked: HouseStanding[];
  unranked: UnrankedHouse[];
  my_ballot: HouseBallot | null;
}

export interface TermStanding {
  chapter_id: string;
  org_name: string;
  chapter_name: string | null;
  weekly_wins: number;
  net: number;
}

export interface TermTitleRace {
  campus_id: string;
  /** "Fall 26" — what gets engraved on the title. */
  term_label: string;
  term_start: string;
  term_end: string;
  weeks_scored: number;
  /** null until a week has actually been won. */
  leader: TermStanding | null;
  standings: TermStanding[];
}

/**
 * Cast or change this week's ballot.
 *
 * PUT because it is idempotent per week by construction: the server keys on
 * (campus, week, voter), so calling this again REPLACES your vote rather than adding
 * one. There is deliberately no week parameter — the server decides which week it is,
 * so a finished week cannot be voted into after the fact.
 */
export async function castHouseBallot(
  campusId: string,
  touseChapterId: string,
  bouseChapterId?: string | null,
): Promise<HouseBallot> {
  return request<HouseBallot>(`/campuses/${campusId}/house-ballot`, {
    method: "PUT",
    body: { touse_chapter_id: touseChapterId, bouse_chapter_id: bouseChapterId ?? null },
  });
}

/** This week's ranking, or a past week's by `weekStart` (read-only addressing). */
export async function getHouseLeaderboard(
  campusId: string,
  weekStart?: string,
): Promise<HouseLeaderboard> {
  return request<HouseLeaderboard>(`/campuses/${campusId}/house-leaderboard`, {
    query: { week_start: weekStart },
  });
}

/** The race for the term title — "Touse of Fall 26". */
export async function getHouseTitleRace(campusId: string): Promise<TermTitleRace> {
  return request<TermTitleRace>(`/campuses/${campusId}/house-title-race`);
}

File: backend/app/schemas/house.py
"""Touse/Bouse weekly house leaderboard schemas (board card c175)."""

import uuid
from datetime import date, datetime

from app.schemas.base import _Schema


class HouseBallotCreate(_Schema):
    """Body for casting or replacing this week's ballot.

    THERE IS NO WEEK FIELD, on purpose. The week is whatever week the server is in when
    the request lands; accepting one would let a client vote into a finished week.

    bouse_chapter_id is optional: naming a top house without naming a bottom one is a
    complete ballot. Requiring both would manufacture Bouse votes from people who did
    not want to cast one, which is the fastest way to make the bottom of the leaderboard
    meaningless.
    """

    touse_chapter_id: uuid.UUID
    bouse_chapter_id: uuid.UUID | None = None


class HouseBallotOut(_Schema):
    """A ballot as its own voter sees it.

    NO voter_id, ever - not even back to the person who cast it. The column exists so
    abuse can be investigated and is never serialized by any route, the same rule
    Chirp.author_id follows (SPEC 8.3). Serializing it "only for the owner" is how a field
    ends up in a shared component that later renders for someone else.
    """

    campus_id: uuid.UUID
    week_start: date
    touse_chapter_id: uuid.UUID
    bouse_chapter_id: uuid.UUID | None = None
    updated_at: datetime


class HouseStanding(_Schema):
    """One house's placing for one week."""

    chapter_id: uuid.UUID
    org_name: str
    chapter_name: str | None = None
    rank: int
    touse_votes: int
    bouse_votes: int
    net: int


class UnrankedHouse(_Schema):
    """A house that exists but has not cleared the ballot threshold this week.

    Reported separately rather than ranked last. Being last and having three votes are
    completely different statements about a real organisation, and collapsing them is
    how a house gets called the worst on campus by a handful of people.
    """

    chapter_id: uuid.UUID
    org_name: str
    chapter_name: str | None = None
    votes: int


class HouseLeaderboard(_Schema):
    """One week's ranking.

    `ballots_cast` is the sample the whole thing is drawn from and is always reported,
    for the same reason the Secretary dashboard always captions its meeting count: a
    ranking with no denominator invites people to read noise as a result.
    """

    campus_id: uuid.UUID
    week_start: date
    ballots_cast: int
    min_votes_to_rank: int
    ranked: list[HouseStanding]
    unranked: list[UnrankedHouse]
    my_ballot: HouseBallotOut | None = None


class TermStanding(_Schema):
    """One house's season record across the term."""

    chapter_id: uuid.UUID
    org_name: str
    chapter_name: str | None = None
    weekly_wins: int
    net: int


class TermTitleRace(_Schema):
    """The race for "Touse of Fall 26".

    `leader` is whoever currently holds the most weekly wins; it is None until at least
    one week has produced a ranked result. The title is decided on WEEKLY WINS rather
    than cumulative points, so one enormous week cannot buy a semester - flagged on c175
    as a judgement call rather than a stated requirement.
    """

    campus_id: uuid.UUID
    term_label: str
    term_start: date
    term_end: date
    weeks_scored: int
    leader: TermStanding | None = None
    standings: list[TermStanding]
